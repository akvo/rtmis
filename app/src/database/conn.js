import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';

const openDatabase = () => {
  const db = SQLite.openDatabase('db.db');
  return db;
};

const init = openDatabase();

const tx = (db, query, params = []) => {
  return new Promise((resolve, reject) => {
    db.transaction(
      (transaction) => {
        if (Array.isArray(query)) {
          const promises = query.map((q) => {
            return new Promise((innerResolve, innerReject) => {
              transaction.executeSql(
                q,
                params,
                (_, resultSet) => {
                  innerResolve(resultSet);
                },
                (_, error) => {
                  innerReject(error);
                  return false; // Rollback the transaction
                },
              );
            });
          });

          Promise.all(promises)
            .then((results) => {
              resolve(results);
            })
            .catch((error) => {
              reject(error);
            });
        } else {
          transaction.executeSql(
            query,
            params,
            (_, resultSet) => {
              resolve(resultSet);
            },
            (_, error) => {
              reject(error);
              return false; // Rollback the transaction
            },
          );
        }
      },
      (error) => {
        reject(error);
      },
    );
  });
};

const openDBfile = async (databaseFile, databaseName) => {
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + 'SQLite')).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + 'SQLite');
  }
  await FileSystem.downloadAsync(
    Asset.fromModule(databaseFile).uri,
    FileSystem.documentDirectory + `SQLite/${databaseName}.db`,
  );
  return SQLite.openDatabase(`${databaseName}.db`);
};

export const conn = {
  file: (dbFile, dbName) => openDBfile(dbFile, dbName),
  init,
  tx,
};
