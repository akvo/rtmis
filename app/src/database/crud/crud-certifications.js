import { conn, query } from '..';

const db = conn.init;
const TABLE_NAME = 'certifications';

const certificationQuery = () => ({
  syncForm: async ({ formId, administrationId, formJSON }) => {
    const findQuery = query.read(TABLE_NAME, { uuid: formJSON.uuid });
    const { rows } = await conn.tx(db, findQuery, [formJSON.uuid]);

    let params = [];
    let queryText = query.insert(TABLE_NAME, {
      formId,
      uuid: formJSON.uuid,
      name: formJSON?.datapoint_name || null,
      administrationId,
      json: formJSON?.answers ? JSON.stringify(formJSON.answers).replace(/'/g, "''") : null,
      syncedAt: new Date().toISOString(),
    });

    if (rows.length) {
      queryText = query.update(
        TABLE_NAME,
        { id: rows._array[0].id },
        {
          json: formJSON?.answers
            ? JSON.stringify(formJSON.answers).replace(/'/g, "''")
            : rows._array[0].json,
        },
      );
      params = [rows._array[0].id];
    }
    const res = await conn.tx(db, queryText, params);
    return res;
  },
  getTotal: async (formId, search, administrationId) => {
    let querySQL = search.length
      ? `SELECT COUNT(*) AS count FROM ${TABLE_NAME} where formId = ? AND name LIKE ? COLLATE NOCASE `
      : `SELECT COUNT(*) AS count FROM ${TABLE_NAME} where formId = ? `;
    const params = search.length ? [formId, `%${search}%`] : [formId];
    if (administrationId) {
      querySQL += ' AND administrationId = ? ';
      params.push(administrationId);
    }
    const { rows } = await conn.tx(db, querySQL, params);
    return rows._array?.[0]?.count;
  },
  getPagination: async ({
    formId,
    search = '',
    limit = 10,
    offset = 0,
    administrationId = null,
  }) => {
    let sqlQuery = `SELECT * FROM ${TABLE_NAME} WHERE formId = $1`;
    const queryParams = [formId];

    if (search.trim() !== '') {
      sqlQuery += ' AND name LIKE $2 COLLATE NOCASE';
      queryParams.push(`%${search}%`);
    }

    if (administrationId) {
      sqlQuery += ' AND administrationId = $3';
      queryParams.push(administrationId);
    }

    sqlQuery += ' ORDER BY syncedAt DESC LIMIT $4 OFFSET $5';
    queryParams.push(limit, offset * limit);
    const { rows } = await conn.tx(db, sqlQuery, queryParams);

    if (!rows.length) {
      return [];
    }
    return rows._array;
  },
  updateIsCertified: async (formId, uuid) => {
    try {
      const updateQuery = query.update(TABLE_NAME, { formId, uuid }, { isCertified: 1 });
      return await conn.tx(db, updateQuery, [formId, uuid]);
    } catch {
      return null;
    }
  },
});

const crudCertification = certificationQuery();

export default crudCertification;
