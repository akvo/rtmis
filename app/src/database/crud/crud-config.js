import { conn, query } from '../';
import defaultBuildParams from '../../build';

const db = conn.init;

const configQuery = () => {
  const id = 1;
  return {
    getConfig: async () => {
      const { rows } = await conn.tx(db, query.read('config', { id }), [id]);
      if (!rows.length) {
        return false;
      }
      return rows._array[rows.length - 1];
    },
    addConfig: async (data = {}) => {
      const insertQuery = query.insert('config', {
        id,
        appVersion: defaultBuildParams.appVersion,
        ...data,
      });
      return await conn.tx(db, insertQuery, []);
    },
    updateConfig: async (data) => {
      const updateQuery = query.update('config', { id }, { ...data });
      return await conn.tx(db, updateQuery, [id]);
    },
  };
};

const crudConfig = configQuery();

export default crudConfig;
