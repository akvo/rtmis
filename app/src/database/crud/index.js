export { default as crudSessions } from './crud-sessions';
export { default as crudForms } from './crud-forms';
export { default as crudUsers } from './crud-users';
export { default as crudConfig } from './crud-config';
export { default as crudDataPoints } from './crud-datapoints';
