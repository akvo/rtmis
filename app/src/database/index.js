export * from './conn';
export * from './query';
export * from './tables';
