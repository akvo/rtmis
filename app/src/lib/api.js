import axios from 'axios';

export const config = {
  baseURL: null,
  headers: {
    'Content-Type': 'application/json',
  },
};

const API = () => {
  const getConfig = () => {
    let current = config;
    if (api.baseURL) {
      current = {
        ...config,
        baseURL: api.baseURL,
      };
    }
    return api?.token
      ? {
          ...current,
          headers: { ...config.headers, Authorization: `Bearer ${api.token}` },
        }
      : current;
  };
  return {
    get: (url, config = {}) => axios({ url, ...getConfig(), ...config }),
    post: (url, data, config = {}) =>
      axios({
        url,
        method: 'POST',
        data,
        ...{
          ...getConfig(),
          headers: {
            ...getConfig().headers,
            ...config?.headers,
          },
        },
      }),
    put: (url, data, config) => axios({ url, method: 'PUT', data, ...getConfig(), ...config }),
    patch: (url, data, config) => axios({ url, method: 'PATCH', data, ...getConfig(), ...config }),
    delete: (url) => axios({ url, method: 'DELETE', ...getConfig() }),
    setToken: (token) => {
      api.token = token;
    },
    setServerURL: (serverURL) => {
      api.baseURL = serverURL;
    },
    getConfig,
  };
};

const api = API();

export default api;
