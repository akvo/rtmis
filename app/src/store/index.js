export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
