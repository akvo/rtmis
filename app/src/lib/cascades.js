import { Asset } from 'expo-asset';
import * as FileSystem from 'expo-file-system';
import * as SQLite from 'expo-sqlite';
import { conn, query } from '../database';

const DIR_NAME = 'SQLite';

const createSqliteDir = async () => {
  /**
   * Setup sqlite directory to save cascades sqlite from server
   */
  if (!(await FileSystem.getInfoAsync(FileSystem.documentDirectory + DIR_NAME)).exists) {
    await FileSystem.makeDirectoryAsync(FileSystem.documentDirectory + DIR_NAME);
  }
};

const download = (downloadUrl, fileUrl) => {
  const fileSql = fileUrl?.split('/')?.pop(); // get last segment as filename
  const pathSql = `${DIR_NAME}/${fileSql}`;
  FileSystem.getInfoAsync(FileSystem.documentDirectory + pathSql).then(({ exists }) => {
    if (!exists) {
      FileSystem.downloadAsync(downloadUrl, FileSystem.documentDirectory + pathSql);
    }
  });
};

const loadDataSource = async (source, id = null) => {
  const { file: cascadeName } = source;
  const db = SQLite.openDatabase(cascadeName);
  const readQuery = id ? query.read('nodes', { id }) : query.read('nodes');
  const readParam = id ? [id] : [];
  const result = await conn.tx(db, readQuery, readParam);
  return result;
};

const dropFiles = async () => {
  const Sqlfiles = await FileSystem.readDirectoryAsync(FileSystem.documentDirectory + DIR_NAME);
  Sqlfiles.forEach(async (file) => {
    if (file.includes('sqlite')) {
      const fileUri = FileSystem.documentDirectory + `${DIR_NAME}/${file}`;
      await FileSystem.deleteAsync(fileUri);
    }
  });
  return Sqlfiles;
};

export default cascades = {
  createSqliteDir,
  loadDataSource,
  download,
  dropFiles,
  DIR_NAME,
};
