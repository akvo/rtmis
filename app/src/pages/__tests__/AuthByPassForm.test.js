import React from 'react';
import renderer from 'react-test-renderer';
import AuthByPassFormPage from '../AuthByPassForm';
import api from '../../lib/api';
import cascades from '../../lib/cascades';
import { UIState, UserState } from '../../store';
import { render, fireEvent, act, renderHook, waitFor } from '@testing-library/react-native';
import { Platform } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { crudUsers } from '../../database/crud';

jest.mock('../../lib/api');
jest.mock('../../database/crud');
jest.mock('../../lib/cascades');
// mock console error
global.console.error = jest.fn();

describe('AuthByPassForm', () => {
  test('it renders correctly', () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const tree = renderer.create(<AuthByPassFormPage navigation={navigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it should have form id input and download button', () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    const fidInput = wrapper.getByTestId('input-form-id');
    expect(fidInput).toBeDefined();
    const downloadButton = wrapper.getByTestId('button-download-form');
    expect(downloadButton).toBeDefined();
  });

  it('it should not download forms when offline', async () => {
    Platform.OS = 'android';
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.post.mockImplementation(() => Promise.resolve({ data: { formsUrl: [] } }));

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = false;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    expect(api.get).not.toHaveBeenCalled();
  });

  it('it should download forms when online', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() => Promise.resolve({ data: { id: 1, name: 'Test' } }));

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    expect(api.get).toHaveBeenCalledWith('/forms/1');
  });

  it('it should navigate to add user if no user defined after form downloaded', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    // url: /forms/1
    const FormId1 = {
      id: 1,
      name: 'Household',
      version: '1.0.0',
      cascades: ['/cascades/1.sqlite'],
      question_group: [],
    };
    // url: /sqlite/file.sqlite
    const mockFile = 'file.sqlite';

    api.getConfig.mockImplementation(() => ({ baseURL: 'http://example.com' }));
    api.get.mockImplementation((url) => {
      if (url === '/forms/1') {
        return Promise.resolve({ data: FormId1 });
      }
      if (url === '/cascades/1.sqlite') {
        return Promise.resolve({ data: mockFile });
      }
    });

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
      expect(cascades.download).toHaveBeenCalledWith(
        'http://example.com/cascades/1.sqlite',
        '/cascades/1.sqlite',
      );
    });
    await waitFor(() => expect(navigation.navigate).toHaveBeenCalledWith('AddUser'));
  });

  it('it should navigate to home if user defined after form downloaded', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    const mockGetData = {
      id: 1,
      name: 'Household',
      version: '1.0.0',
      cascades: ['/cascades/1.sqlite'],
      question_group: [],
    };
    const mockUser = { id: 1, name: 'John Doe', password: 'qwerty' };
    api.get.mockImplementation(() => Promise.resolve({ data: mockGetData }));
    crudUsers.getActiveUser.mockImplementation(() => Promise.resolve(mockUser));
    const { result: userStateRef } = renderHook(() => UserState.useState((s) => s));

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    expect(api.get).toHaveBeenCalledWith('/forms/1');
    await waitFor(() => {
      const {
        id: userIdState,
        name: userNameState,
        password: userPasswordState,
      } = userStateRef.current;
      expect(userIdState).toEqual(mockUser.id);
      expect(userNameState).toEqual(mockUser.name);
      expect(userPasswordState).toEqual(mockUser.password);
      expect(navigation.navigate).toHaveBeenCalledWith('Home');
    });
  });

  it('it should be error 400', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() =>
      Promise.reject({ response: { message: 'Failed', status: 400 } }),
    );

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
      const errorText = wrapper.getByTestId('fetch-error-text');
      expect(errorText).toBeTruthy();
    });
  });

  it('it should be error 500', async () => {
    const { result: navigationRef } = renderHook(() => useNavigation());
    const navigation = navigationRef.current;
    api.get.mockImplementation(() =>
      Promise.reject({ response: { message: 'Failed', status: 500 } }),
    );

    const wrapper = render(<AuthByPassFormPage navigation={navigation} />);

    act(() => {
      UIState.update((s) => {
        s.online = true;
      });
    });

    const fidInput = wrapper.getByTestId('input-form-id');
    const downloadButton = wrapper.getByTestId('button-download-form');

    fireEvent.changeText(fidInput, '1');
    fireEvent.press(downloadButton);

    await waitFor(() => {
      expect(api.get).toHaveBeenCalledWith('/forms/1');
    });
  });
});
