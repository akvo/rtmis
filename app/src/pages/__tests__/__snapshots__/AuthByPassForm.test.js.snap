// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthByPassForm it renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#f9fafb",
      "display": "flex",
      "flex": 1,
      "flexDirection": "column",
      "gap": 36,
      "justifyContent": "center",
      "paddingHorizontal": 16,
    }
  }
  testID="center-layout"
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      {
        "backgroundColor": "#e5e7eb",
        "borderRadius": 4,
        "height": 110,
        "overflow": "hidden",
        "position": "relative",
        "width": 110,
      }
    }
  >
    <Image
      onLoad={[Function]}
      source={
        {
          "uri": "uri",
        }
      }
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="image-component"
      theme={
        {
          "colors": {
            "background": "#ffffff",
            "black": "#242424",
            "disabled": "hsl(208, 8%, 90%)",
            "divider": "#bcbbc1",
            "error": "#ff190c",
            "grey0": "#393e42",
            "grey1": "#43484d",
            "grey2": "#5e6977",
            "grey3": "#86939e",
            "grey4": "#bdc6cf",
            "grey5": "#e1e8ee",
            "greyOutline": "#bbb",
            "platform": {
              "android": {
                "error": "#f44336",
                "grey": "rgba(0, 0, 0, 0.54)",
                "primary": "#2196f3",
                "searchBg": "#dcdce1",
                "secondary": "#9C27B0",
                "success": "#4caf50",
                "warning": "#ffeb3b",
              },
              "default": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "ios": {
                "error": "#ff3b30",
                "grey": "#7d7d7d",
                "primary": "#007aff",
                "searchBg": "#dcdce1",
                "secondary": "#5856d6",
                "success": "#4cd964",
                "warning": "#ffcc00",
              },
              "web": {
                "error": "#ff190c",
                "grey": "#393e42",
                "primary": "#2089dc",
                "searchBg": "#303337",
                "secondary": "#ca71eb",
                "success": "#52c41a",
                "warning": "#faad14",
              },
            },
            "primary": "#2089dc",
            "searchBg": "#303337",
            "secondary": "#ad1457",
            "success": "#52c41a",
            "warning": "#faad14",
            "white": "#ffffff",
          },
          "spacing": {
            "lg": 12,
            "md": 8,
            "sm": 4,
            "xl": 24,
            "xs": 2,
          },
        }
      }
      transitionDuration={360}
    />
    <View
      accessibilityElementsHidden={true}
      collapsable={false}
      importantForAccessibility="no-hide-descendants"
      pointerEvents="none"
      style={
        {
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#bdbdbd",
            "justifyContent": "center",
          }
        }
        testID="RNE__Image__placeholder"
      />
    </View>
    <View
      style={{}}
      testID="RNE__Image__children__container"
    />
  </View>
  <View>
    <View
      style={
        {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": "75%",
          }
        }
      >
        <TextInput
          autoFocus={true}
          editable={true}
          keyboardType="numeric"
          onChangeText={[Function]}
          placeholder="Form ID"
          placeholderTextColor="#86939e"
          style={
            {
              "color": "#242424",
              "flex": 1,
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="input-form-id"
          underlineColorAndroid="transparent"
          value={null}
        />
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          {
            "busy": false,
            "checked": undefined,
            "disabled": false,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
        testID="button-download-form"
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#2089dc",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text
            style={
              {
                "color": "white",
                "fontSize": 18,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Download Form
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
